### ISSUE TITLE
CLI: Refactor using context-object approach

### ISSUE THREAD

**Issue by Makman2** [coala/coala#4114]
From [this discussion on gitter](https://gitter.im/coala/coala?at=58f39ac84cb8d09173975349):

Currently everything is quite cluttered, and we have many many unintuitive cross-references to functions or objects. Together with passing around log-printers this makes the current CLI a mess from code-perspective, that's why I think that we should implement a "context-object"-based approach.

i.e.

instead of passing thousands of parameters around to each function that might need something (because one of the subfunctions needs something), we use a `coalaRunContext`. When coala gets started, we instantiate a new `coalaRunContext`, where we feed in stuff like passed arguments to coala, maybe sections, etc (what is taken over by the context and what not needs consideration).

Actually it is from designer perspective not the perfect approach in fact, I'm generally not a fan of context objects (as in our case it's in fact a singleton, singletons shouldn't be objects, and in python you can avoid them because functions are objects. That's why we ended up with our current design ;D).

But in this case it will help I believe. We do cross-reference so many data, so it would also speed up things and maintain readability.

CC @satwikkansal

**Comment by satwikkansal** [comment]
Another related discussion: https://gitter.im/coala/coala?at=58f3b84ca0e48562420dcea0

**Comment by satwikkansal** [comment]
I was wondering if we can 
- Have a single entry point for every module (which we already have)
- Initialize a context using something like 
  ```py
  context = dict(**locals())
  ```
- Make this context available to all other methods in the module.
- Other methods can update the values of the variables in the context, but cannot add/delete a new variable.

This way we could reduce a lot of function signatures and inline documentation, since we'll have to do it only at the entry point method. 
We can further extend this to a coalaContext by not limiting the scope of context to module level, but I think module-level context will be easier to begin with.

**Comment by Makman2** [comment]
Yeah maybe module level is easier. Though I would prefer that we end with a class because saving data in modules is nearly never good^^
